docs/proposals/parameterize-strategies.md

```diff
@@ -0,0 +1,267 @@
+
+- Introduce a `spec.params`on Strategies, both cluster or namespaced scope. This allow strategy administrators to layout the definition of N parameters, as required for their strategies.
+
+- Revisit what we consider today as default Parameters that are defined on runtime when creating `TaskRuns`. These are `DOCKERFILE`, `CONTEXT_DIR` and `BUILDER_IMAGE`. Decide if we need to reduce or extend the default definitions.
+
+- Expose the usage of Tekton Params directly on the strategies, in favor of simplicity and readability. This boils down to `inputs.params.DOCKERFILE`, instead of `build.dockerfile`.
+
+### Non-Goals
+
+- Build from scratch a paremeterize mechanism, by replacing the one available in Task/TaskRuns.
+
+- Provide support for Tekton Params of the type `array`.
+
+## Proposal
+
+### Part 1: Introduce spec.params
+
+Introduce the `spec.params` field across Build, BuildRun and BuildStrategies resources. This will define a list  of parameters definition, of the type `string`. This new `spec.params` does not provide support for Tekton params of the type `array`. This field can only be use, if the parameter is supported in the strategy of choice.
+
+For example:
+
+```yaml
+apiVersion: shipwright.io/v1alpha1
+kind: ClusterBuildStrategy
+metadata:
+  name: a-cluster-strategy
+spec:
+  buildSteps: #Content omitted for this example
+  params:
+  - name: a-param
+    description: A description of this parameter definition.
+    default: The default parameter value.
+```
+
+```yaml
+apiVersion: shipwright.io/v1alpha1
+kind: Build
+metadata:
+  name: a-build
+spec:
+  source: #Content omitted for this example
+  strategy:
+    name: a-strategy-with-params
+  output: #Content omitted for this example
+  params:
+  - name: a-param
+    value: A parameter value.
+```
+
+```yaml
+apiVersion: shipwright.io/v1alpha1
+kind: BuildRun
+metadata:
+  name: a-buildrun
+spec:
+  buildRef:
+    name: a-build
+  params:
+  - name: a-param
+    value: Another parameter value because my build is not up-to-date.
+```
+
+_Note_: If a **Buildrun** specifies `a-param` via its `spec.params` ,this will override the value defined in the `a-build`. In other words, BuildRuns have a higher priority when defining parameters.
+
+### Part 2: Runtime Parameters
+
+As mentioned in the Goals section, we currently define three parameters that can be constructed on runtime, also know as _runtime parameters_. Runtime parameters are define during runtime, on the creation of a Taskrun, where the paremeter definition and the parameter value mapping takes place. These runtime parameters are:
+
+- DOCKERFILE
+- CONTEXT_DIR
+- BUILDER_IMAGE (_optional: It only take place if the spec.builder.image is defined_)
+
+This EP propose to stop using `BUILDER_IMAGE` as a runtime Parameter but rather to delegate its definition to user of N strategy. This means `BUILDER_IMAGE` should be defined under `spec.params` in the future.
```

Comment:
I see, not sure, definitely not as part of this EP, but we might need an issue on that, because it seems even on the current state we don't have a way to specify a secret, not saying we must.

Reply:
Yeah, I think we can defer it. Would maybe end up as:

Strategy:

```yaml
  params:
    - name: builder-image
      description: The builder image.
      type: string
    - name: builder-image-secret
      description: A secret to pull the builder image.
      type: secret
  volumes:
  - name: builder-image-secret
    secret:
      secretName: $(builder-image-secret)
  buildSteps:
    - command:
        - /usr/local/bin/s2i
        - build
        - .
        - $(params.builder-image)
        - '--as-dockerfile'
        - /gen-source/Dockerfile.gen
      image: 'quay.io/openshift-pipeline/s2i:nightly'
      name: s2i-build-as-dockerfile
      volumeMounts:
        - mountPath: /gen-source
          name: gen-source
        - volumeMounts:
          - name: builder-image-secret
             mountPath: "/somePathForBuilderImageSecret"
             readOnly: true
      workingDir: /workspace/source
```